config/init_server_creds.go

```diff
@@ -503,44 +567,186 @@ func GenerateCert() error {
-	// Check to see if we already had an IssuerKey or generate one
-	if err := GeneratePrivateKey(issuerKeyFile, elliptic.P256(), false); err != nil {
-		return nil, errors.Wrap(err, "Failed to generate new private key")
-	}
-	contents, err := os.ReadFile(issuerKeyFile)
+// Helper function to initialize the in-memory map to save all private keys
+func initKeysMap() {
+	initialMap := make(map[string]jwk.Key)
+	issuerKeys.Store(&IssuerKeys{
+		AllKeys: initialMap,
+	})
+}
+
+// Helper function to load one .pem file from specified filename
+func loadSinglePEM(path string) (jwk.Key, error) {
+	contents, err := os.ReadFile(path)
 	if err != nil {
-		return nil, errors.Wrap(err, "Failed to read issuer key file")
+		return nil, errors.Wrap(err, "failed to read key file")
 	}
+
 	key, err := jwk.ParseKey(contents, jwk.WithPEM(true))
 	if err != nil {
-		return nil, errors.Wrapf(err, "Failed to parse issuer key file %v", issuerKeyFile)
+		return nil, errors.Wrapf(err, "failed to parse issuer key file %v", path)
 	}
 
 	// Add the algorithm to the key, needed for verifying tokens elsewhere
-	err = key.Set(jwk.AlgorithmKey, jwa.ES256)
+	if err := key.Set(jwk.AlgorithmKey, jwa.ES256); err != nil {
+		return nil, errors.Wrap(err, "failed to set algorithm")
+	}
+
+	// Ensure key has an ID
+	if err := jwk.AssignKeyID(key); err != nil {
+		return nil, errors.Wrap(err, "failed to assign key ID")
+	}
+
+	return key, nil
+}
+
+// Helper function to load/refresh all key files from both legacy IssuerKey file and specified directory
+// find the most recent private key based on lexicographical order of their filenames
+func loadPEMFiles(dir string) (jwk.Key, error) {
+	var mostRecentKey jwk.Key
+	var mostRecentFileName string
+	latestKeys := getIssuerPrivateKeysCopy()
+
+	// Load legacy private key if it exists - parsing the file at IssuerKey act as if it is included in IssuerKeysDirectory
+	issuerKeyPath := param.IssuerKey.GetString()
+	if issuerKeyPath != "" {
+		if _, err := os.Stat(issuerKeyPath); err == nil {
+			issuerKey, err := loadSinglePEM(issuerKeyPath)
+			if err != nil {
+				log.Warnf("Failed to load key %s: %v", issuerKeyPath, err)
+			} else {
+				latestKeys[issuerKey.KeyID()] = issuerKey
+				if mostRecentFileName == "" || filepath.Base(issuerKeyPath) > mostRecentFileName {
+					mostRecentFileName = filepath.Base(issuerKeyPath)
+					mostRecentKey = issuerKey
+				}
+			}
+		}
+	}
+
+	// Ensure input directory dir exists, if not, create it with proper permissions
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := createDirForKeys(dir); err != nil {
+			return nil, errors.Wrapf(err, "failed to create directory and set permissions: %s", dir)
+		}
+	}
+	// Traverse the directory for .pem files in lexical order
+	err := filepath.WalkDir(dir, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !d.IsDir() && filepath.Ext(d.Name()) == ".pem" {
```

Comment:
Two items:
- I want to avoid single-character variable names.  Humans struggle with understanding them from context.  something like `dirEntry` is much better than `d`.
- Use [`SkipDir`](https://pkg.go.dev/io/fs#SkipDir) to avoid recursing into a subdirectory.